File: Algorithms/Python/pattern_searching/kmp.py
from __future__ import annotations


def kmp_search(text: str, pattern: str) -> int:
    n, m = len(text), len(pattern)
    if m == 0:
        return 0
    lps = [0] * m
    compute_lps(pattern, m, lps)
    i, j = 0, 0
    while i < n:
        if text[i] == pattern[j]:
            i += 1
            j += 1
        if j == m:
            return i - j
        elif i < n and text[i] != pattern[j]:
            if j != 0:
                j = lps[j - 1]
            else:
                i += 1
    return -1


def compute_lps(pattern: str, m: int, lps: list[int]) -> None:
    length, i = 0, 1
    while i < m:
        if pattern[i] == pattern[length]:
            length += 1
            lps[i] = length
            i += 1
        elif length == 0:
            lps[i] = 0
            i += 1

        else:
            length = lps[length - 1]


if __name__ == "__main__":
    text = "CAABAACAADAABAABA"
    pattern = "AABA"
    index = kmp_search(text, pattern)
    if index != -1:
        print(f"Pattern found at index {index}")
    else:
        print("Pattern not found")

File: Algorithms/Javascript/pattern_searching/kmp.js
const computeLPSArray = (pattern) => {
    const m = pattern.length;
    const lps = new Array(m).fill(0);

    let len = 0;
    let i = 1;

    while (i < m) {
        if (pattern[i] === pattern[len]) {
            len++;
            lps[i] = len;
            i++;
        }
        else if (len !== 0) {
            len = lps[len - 1];
        }
        else {
            lps[i] = 0;
            i++;
        }
    }

    return lps;
}

const kmp = (text, pattern) => {
    const n = text.length;
    const m = pattern.length;
    const lps = computeLPSArray(pattern);

    let i = 0;
    let j = 0;

    while (i < n) {
        if (text[i] === pattern[j]) {
            i++;
            j++;
        }

        if (j === m) {
            console.log(`Pattern found at index ${i - j}`);
            j = lps[j - 1];
        } else if (i < n && text[i] !== pattern[j]) {
            if (j !== 0) {
                j = lps[j - 1];
            } else {
                i++;
            }
        }
    }
}

const text = 'AABAACAADAABAAABAA';
const pattern = 'AABA';

kmp(text, pattern);
